runtime/interpreter/value_range.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	step := GetValueForIntegerType(1, rangeStaticType.ElementType)
+	if startComparable.Greater(interpreter, endComparable, locationRange) {
+		elemSemaTy := interpreter.MustConvertStaticToSemaType(rangeStaticType.ElementType)
+		if elemSemaTy.Tag().BelongsTo(sema.UnsignedIntegerTypeTag) {
+			panic(InclusiveRangeConstructionError{
+				LocationRange: locationRange,
+				Message: fmt.Sprintf(
+					"step value cannot be negative for unsigned integer type %s",
+					elemSemaTy,
+				),
+			})
+		}
+
+		negatedStep, ok := step.Negate(interpreter, locationRange).(IntegerValue)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+
+		step = negatedStep
+	}
+
+	return createInclusiveRange(
+		interpreter,
+		locationRange,
+		start,
+		end,
+		step,
+		rangeStaticType,
+		rangeSemaType,
+	)
+}
+
+// NewInclusiveRangeValue constructs an InclusiveRange value with the provided start, end & step.
+// NOTE: Assumes that the values start, end and step are of the same static type.
+func NewInclusiveRangeValueWithStep(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	start IntegerValue,
+	end IntegerValue,
+	step IntegerValue,
+	rangeType InclusiveRangeStaticType,
+	rangeSemaType *sema.InclusiveRangeType,
+) *CompositeValue {
+
+	zeroValue := GetValueForIntegerType(0, start.StaticType(interpreter))
+
+	// Validate that the step is non-zero.
+	if step.Equal(interpreter, locationRange, zeroValue) {
+		panic(InclusiveRangeConstructionError{
+			LocationRange: locationRange,
+			Message:       "step value cannot be zero",
+		})
+	}
+
+	// Validate that the sequence is moving towards the end value.
+	// If start < end, step must be > 0
+	// If start > end, step must be < 0
+	// If start == end, step doesn't matter.
+	if isSequenceMovingAwayFromEnd(interpreter, locationRange, start, end, step, zeroValue) {
+
+		panic(InclusiveRangeConstructionError{
+			LocationRange: locationRange,
+			Message: fmt.Sprintf(
+				"sequence is moving away from end: %s due to the value of step: %s and start: %s",
+				end,
+				step,
+				start,
+			),
+		})
+	}
+
+	return createInclusiveRange(
+		interpreter,
+		locationRange,
+		start,
+		end,
+		step,
+		rangeType,
+		rangeSemaType,
+	)
+}
+
+func createInclusiveRange(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	start IntegerValue,
+	end IntegerValue,
+	step IntegerValue,
+	rangeType InclusiveRangeStaticType,
+	rangeSemaType *sema.InclusiveRangeType,
+) *CompositeValue {
+	fields := []CompositeField{
+		{
+			Name:  sema.InclusiveRangeTypeStartFieldName,
+			Value: start,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeEndFieldName,
+			Value: end,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeStepFieldName,
+			Value: step,
+		},
+	}
+
+	rangeValue := NewCompositeValueWithStaticType(
+		interpreter,
+		locationRange,
+		nil,
+		rangeSemaType.QualifiedString(),
+		common.CompositeKindStructure,
+		fields,
+		common.ZeroAddress,
+		rangeType,
+	)
+
+	rangeValue.Functions = map[string]FunctionValue{
```

Comment:
The functions must also be injected when the value is loaded from storage.

Could you please add a `runtime` test to ensure this works?
Basically, performs two transactions:
- Create range and save it to account storage
- Load range from storage and call `contains`

We'll need to see how we can inject it, given the value is/us stored as a `CompositeValue`

Reply:
It works. Added a case in `TestInterpretAuthAccount_load`.

Reply:
The added testcase isn't quite testing that yet, see https://github.com/onflow/cadence/pull/2523/files#r1344848593. 

Add a runtime test with two transactions. It will likely fail, because the function isn't linked in after decoding from storage (or I can't quite see how it is at the moment)

Reply:
Sorry, just forgot to put the final comment in this thread. We can't do this since InclusiveRange cannot be stored in account storage.

Filed https://github.com/onflow/cadence/issues/2870 for this.